Significant rewrite

kdawgsf opened the pull request:
- Split dp donor and student out to separate concepts
- Instead of tracking changes, make changes in-place and then compare at the end
- Add email update logic based on latest requirements
- Move instructions to the end of the output
- Output csv files now use appropriate subsets of report 271 fields
- Add more fields to report 271

Commit:
Significant rewrite

- Split dp donor and student out to separate concepts
- Instead of tracking changes, make changes in-place and then compare at the end
- Add email update logic based on latest requirements
- Move instructions to the end of the output
- Output csv files now use appropriate subsets of report 271 fields
- Add more fields to report 271

Commit:
Add the logic to check/uncheck NOMAIL depending on NOBSD status

Commit:
Rename script and move helper functions to utils.py

arvindraghavan commented on district_data_import.py:
```diff
@@ -0,0 +1,415 @@
+            'ZIP': district_record['zip']
+        })
+
+        if district_record['Parent1Email'] and not dp_donorrecord['EMAIL']:
```
Need a detailed comment before this line explaining what requirement or logic we are implementing and any caveats.

arvindraghavan commented on district_data_import.py:
```diff
@@ -0,0 +1,415 @@
+
+    # TODO The else case is multiple records. Need to see if anything changed
+
+    if (district_address not in dp_addresses_by_donor_id.values()) or not dp_emails.issuperset(district_emails):
```
Ditto. We should put a note here to explain to our future selves what requirement we are trying to meet here.

arvindraghavan commented on district_data_import.py:
```diff
@@ -0,0 +1,415 @@
+district_records = {}
+for row in utils.load_csv_file(filename_district_data, DISTRICT_DATA_HEADERS):
+    district_records[row['SystemID']] = row
+
```
I like that we have clean set of populated data structures at this point in the code. However, it has taken me some time to figure out what you are trying to do, and I'm not sure I've fully understood everything. One thing that concerns me is the maintainability of this large monolith. We can make things easier by: 
- Breaking up the code in to several functions. The first function creates the dicts. Then a sequence of five functions should follow that each work with a specific subset of these dicts, and puts out their respective output files. 
- If I understand the logic correctly, we create an unmodified reference, and the two modified dicts, and extract the output files by filtering these modified dicts. If this is not the case, I've totally missed the point, and will need to have you walk me through the logic. I think it is more preferable to keep all the logic pertaining to an output file within a function that simply transforms the provided inputs. Basically refactor towards a functional programming approach. It will allow us to better maintain these functions independently. 
- Finally, each function will need notes/comments explaining the assumptions used to derive the logic.

kdawgsf replied:
You understand correctly: 1) Import csv file into memory, 2) Overlay district data on in-memory structure, 3) Output csv files to bring DP into line with that in-memory structure.

I think you're suggesting that we should not do step 2 but should instead have a series of side-effect-free 2/3 combinations? If that's the case, I don't think so. That's essentially what we had before, and I think it was very limiting. For example, setting the NOMAIL stuff properly requires that you have an up-to-date version of everything in memory in order to count students. Also, this allows us to do sophisticated stuff like filter the fields eligible for modification simply by tweaking step 3. This is very useful for testing.

I think I see how to extract all the DP data to a separate class, which I'll go ahead and do. I'll let you know once I've updated the PR.

arvindraghavan replied:
@kdawgsf (adding this in case comments don't generate e-mails)

In the code as it stands, it seems like the first three output files (student updates, new students, new donors) can be generated right after the the blocks of code that processes the district data for these outputs. I see the NOMAIL logic as part of the existing donor update step (the fourth output). We should be able to craft these four functions without any side effects, and add an additional output from these functions (for example, even a simple list of updated IDs) that can then be processed by a final "validation" function that verifies that the previous functions did their job correctly. This final function should greatly help with testing. 

The current "monolith" design, with very few notes explaining the thinking behind it, will be hard for us to maintain when we look at it again in a few months time. 

I can put more comments in the code explaining where I see the breakouts into functions occurring, in case it is not clear. But the main idea is to generate each output file right after the business logic is applied to the input structures (which includes both DP and BSD data groomed as you already have for easy access). 

We can talk about this on the phone if you prefer.

Commit:
Move to argparse command line parsing

arvindraghavan commented on district_data_import.py:
```diff
@@ -0,0 +1,411 @@
+district_records = {}
+for row in utils.load_csv_file(args.district_data, DISTRICT_DATA_HEADERS):
+    district_records[row['SystemID']] = row
+
```
@kdawgsf, at this point, all the input data has been nicely groomed, and I think we should be able to easily proceed with a functional decomposition.

Commit:
Extract DP-only stuff to its own class, DPData

- Also, finalize guardian email updates and address/email updates for donors
- Also, extract district utils to a separate module

Commit:
Simplify logic for multi-donor email updates

Commit:
Add TODO about matching
